engine/base.ftl

```diff
@@ -1470,3 +1470,373 @@ A comparator is an operator and a version
+Match Filter - filter to be checked.
+
+The "any" behaviour requires that at least one value of the "Any" attribute of the
+Match Filter needs to match one value in any of the attributes of the Current Filter.
+
+Current Filter
+{
+    "Environment" : ["prod"]
+}
+
+matches
+
+Match Filter
+{
+    "Any" : ["prod"]
+}
+
+The "onetoone" behaviour requires that a value of each attribute of the Match Filter
+must match a value of the same named attribute in the Current Filter. The Match
+Filter is thus a subset of the Current Filter.
+
+Current Filter
+{
+    "Product" : ["ics"],
+    "Environment" : ["prod"]
+    "Segment" : ["default"]
+}
+
+matches
+
+Match Filter
+{
+    "Product" : ["ics"],
+    "Environment" : ["prod"],
+    "Segment" : ["e7", "default"]
+}
+
+but the Current Filter
+{
+    "Environment" : ["prod"]
+    "Segment" : ["default"]
+}
+
+does not match.
+
+The "exactonetoone" behaviour uses the same match logic as "onetoone", but
+also checks that the Current Filter doesn't have any attributes that
+are not included in the Match Filter
+
+--]
+
+[#assign ANY_FILTER_MATCH_BEHAVIOUR = "any"]
+[#assign ONETOONE_FILTER_MATCH_BEHAVIOUR = "onetoone"]
+[#assign EXACTLY_ONETOONE_FILTER_MATCH_BEHAVIOUR = "exactlyonetoone"]
+
+[#function isFilterAttribute filter attribute]
+    [#return filter[attribute]??]
+[/#function]
+
+[#function getFilterAttribute filter attribute]
+    [#return asArray(filter[attribute]![]) ]
+[/#function]
+
+[#function getFilterAttributePrimaryValue filter attribute]
+    [#return (getFilterAttribute(filter,attribute)[0])!"" ]
+[/#function]
+
+[#-- Check for a match between a Current Filter and a Match Filter --]
+[#function filterMatch currentFilter matchFilter matchBehaviour]
+
+    [#switch matchBehaviour]
+        [#case ANY_FILTER_MATCH_BEHAVIOUR]
+            [#if !(matchFilter.Any??)]
+                [#return true]
+            [/#if]
+            [#list currentFilter as key, value]
+                [#if getArrayIntersection(value, matchFilter.Any)?has_content]
+                    [#return true]
+                [/#if]
+            [/#list]
+            [#break]
+
+        [#case ONETOONE_FILTER_MATCH_BEHAVIOUR]
+        [#case EXACTLY_ONETOONE_FILTER_MATCH_BEHAVIOUR]
+            [#list matchFilter as key,value]
+                [#if !(currentFilter[key]?has_content)]
+                    [#return false]
+                [/#if]
+                [#if !getArrayIntersection(currentFilter[key],value)?has_content]
+                    [#return false]
+                [/#if]
+            [/#list]
+            [#-- Filters must have the same attributes --]
+            [#if
+                (matchBehaviour == EXACTLY_ONETOONE_FILTER_MATCH_BEHAVIOUR) &&
+                removeObjectAttributes(currentFilter, matchFilter?keys)?has_content]
+                [#return false]
+            [/#if]
+            [#return true]
+            [#break]
+
+            [#-- Unknown behaviour --]
```

Comment:
Perhaps a debug message here to help with troubleshooting when filter isn't being found and you aren't sure why?

Reply:
done - throwing a fatal as we shouldn't ever get here